dialectic-reconnect/src/backoff.rs

```diff
@@ -0,0 +1,99 @@
+use std::time::Duration;
+
+/// A description of a backoff strategy with optional exponential delay, random jitter, maximum
+/// delay, and maximum retries. This can be used to generate retry strategies for building
+/// [`Connector`](crate::retry::Connector)s.
+#[derive(Debug, Clone)]
+pub struct Backoff {
+    initial_delay: Duration,
+    factor: f64,
+    jitter: Duration,
+    max_delay: Option<Duration>,
+    max_retries: usize,
+}
+
+impl Backoff {
+    /// Create a simple [`Backoff`] which delays by `initial_delay` each time it is invoked,
+    /// forever.
+    pub fn with_delay(initial_delay: Duration) -> Self {
+        Backoff {
+            initial_delay,
+            factor: 1.0,
+            jitter: Duration::from_millis(0),
+            max_delay: None,
+            max_retries: usize::MAX,
+        }
+    }
+
+    /// Add an exponential factor to a [`Backoff`], so that every time it is invoked, it delays for
+    /// that multiple of its previous delay time.
+    pub fn exponential(mut self, factor: f64) -> Self {
```

Comment:
All the methods take `mut self` as a value, because I wanted things to be used "builder style" and didn't want to make `Backoff` `Copy` (although it could be, at present).

Reply:
Why does `&mut self` imply that `Backoff` would have to be `Copy`? I'm referring to https://doc.rust-lang.org/1.0.0/style/ownership/builders.html#non-consuming-builders-(preferred): to be clear. i think it would require a lifetime parameter, which is slightly annoying, but it lets you split things up like:

```rust
let backoff = Backoff::new();

if slow_network {
  backoff.exponential(5);
}

backoff.build();
```

Reply:
Ah! A `fn(&mut self) -> Backoff` would need `Backoff: Copy`, but the pattern you linked to has the builder functions as `fn(&mut self) -> &mut Backoff`, which makes total sense. I'll try refactoring to use the suggested pattern in this reference.